File: cpp_ipu_dir/pi.cpp
#include <cstdlib>
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>

#include <poplar/DeviceManager.hpp>
#include <poplar/Device.hpp>
#include <poplar/Target.hpp>
#include <poplar/Graph.hpp>
#include <poplar/Engine.hpp>
#include <poplar/CodeletFileType.hpp>
#include <poplar/Program.hpp>
#include <poplar/Tensor.hpp>

using namespace poplar;
using namespace poplar::program;

int main(int argc, char **argv) {
    auto device_manager = DeviceManager();
    auto max_dev_id = device_manager.getNumDevices();
    Device device;
    for (unsigned id=0; id<max_dev_id; id++) {
        device = device_manager.getDevice(id);
        if (device.attach()) {
            std::cerr << "Using HW device ID: " << device.getId() << "\n";
            break;
        }
    }
    auto target = device.getTarget();
    Graph graph(target);

    auto num_tiles = target.getNumTiles();
    unsigned n = UINT32_MAX / num_tiles;
    unsigned num_steps = num_tiles * n;
    float slice = 1 / num_steps;

    auto tile_clock_frequency = target.getTileClockFrequency();

    unsigned ids[num_tiles];
    for (size_t i=0; i<num_tiles; i++) {
        ids[i] = i;
    }
    std::vector<float> sums(num_tiles);
    std::vector<unsigned> cycles(num_tiles);

    graph.addCodelets({"codelets.cpp"}, "-O3");

    Tensor input = graph.addConstant(UNSIGNED_INT, {num_tiles}, ids);
    Tensor output = graph.addVariable(FLOAT, {num_tiles}, "sums");
    Tensor cs = graph.addVariable(UNSIGNED_INT, {num_tiles}, "cycles");

    Sequence prog;

    auto compute_set = graph.addComputeSet("Pi");
    for (size_t i = 0; i < num_tiles; i++) {
        graph.setTileMapping(input[i], i);
        graph.setTileMapping(output[i], i);
        graph.setTileMapping(cs[i], i);
        auto vtx = graph.addVertex(compute_set, "VertexPi");
        graph.connect(vtx["id"], input[i]);
        graph.connect(vtx["n"], n);
        graph.connect(vtx["slice"], slice);
        graph.connect(vtx["out"], output.slice(i, i+1));
        graph.connect(vtx["cycles"], cs.slice(i, i+1));
        graph.setTileMapping(vtx, i);
    }
    prog.add(program::Execute(compute_set));

    graph.createHostRead("sums-read", output);
    graph.createHostRead("cycles-read", cs);

    Engine engine(graph, prog);
    engine.loadAndRun(device);

    engine.readTensor("sums-read", sums.data(), sums.data() + sums.size());
    engine.readTensor("cycles-read", cycles.data(), cycles.data() + cycles.size());

    device.detach();

    float pi = std::accumulate(sums.begin(), sums.end(), 0);
    pi *= slice;
    auto max_cycles = *std::max_element(cycles.begin(), cycles.end());
    double time = max_cycles / tile_clock_frequency;

    std::cout << "Calculating PI using:" << std::endl;
    std::cout << "  " << num_steps << " slices" << std::endl;
    std::cout << "  " << num_tiles << " IPU tiles" << std::endl;
    std::cout << "Obtained value of PI: " << pi << std::endl;
    std::cout << "Time taken: " << time << " seconds (" << max_cycles << " cycles at " << tile_clock_frequency / 1e9 << " GHz)" << std::endl;

    return 0;
}

File: julia_ipu_pi_dir/pi.jl
#!/usr/bin/env julia

using IPUToolkit.IPUCompiler, IPUToolkit.Poplar

ENV["POPLAR_RUNTIME_OPTIONS"] = """{"target.hostSyncTimeout":"20"}"""

device = Poplar.get_ipu_device()
target = Poplar.DeviceGetTarget(device)
graph = Poplar.Graph(target)

num_tiles = Int(Poplar.TargetGetNumTiles(target))
n::UInt32 = typemax(UInt32) ÷ num_tiles
num_steps::UInt32 = num_tiles * n
slice::Float32 = 1 / num_steps

tile_clock_frequency = Poplar.TargetGetTileClockFrequency(target)

ids = collect(UInt32.(0:(num_tiles - 1)))
sums = similar(ids, Float32)
cycles = similar(ids)

@eval function pi_kernel(i::T) where {T<:Integer}
    sum = 0f0
    for j in (i * $(n)):((i + one(T)) * $(n) - one(T))
        x = (j - 5f-1) * $(slice)
        sum += 4 / (1 + x ^ 2)
    end
    return sum
end

@codelet graph function Pi(in::VertexVector{UInt32, In},
                           out::VertexVector{Float32, Out},
                           cycles::VertexVector{UInt32, Out})
    cycles[begin] = @ipuelapsed(out[begin] = pi_kernel(in[begin]))
end

input = Poplar.GraphAddConstant(graph, ids)
output = similar(graph, input, Float32, "sums");
cs = similar(graph, input, UInt32, "cycles");

prog = Poplar.ProgramSequence()

add_vertex(graph, prog, 0:(num_tiles - 1), Pi, input, output, cs)

Poplar.GraphCreateHostRead(graph, "sums-read", output)
Poplar.GraphCreateHostRead(graph, "cycles-read", cs)

engine = Poplar.Engine(graph, prog)
Poplar.EngineLoadAndRun(engine, device)
Poplar.EngineReadTensor(engine, "sums-read", sums)
Poplar.EngineReadTensor(engine, "cycles-read", cycles)

Poplar.DeviceDetach(device)

pi = sum(sums) * slice
time = round(maximum(cycles) / tile_clock_frequency; sigdigits=4)

print("""
      Calculating PI using:
        $(num_steps) slices
        $(num_tiles) IPU tiles
      Obtained value of PI: $(pi)
      Time taken: $(time) seconds ($(maximum(cycles)) cycles at $(round(tile_clock_frequency / 1e9; sigdigits=3)) GHz)
      """)
